File: src/main/java/br/com/controle/financeiro/controller/RestResponseEntityExceptionHandler.java
package br.com.controle.financeiro.controller;

import br.com.controle.financeiro.model.exception.NotFoundException;

import org.springframework.dao.EmptyResultDataAccessException;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

@RestControllerAdvice
public class RestResponseEntityExceptionHandler extends ResponseEntityExceptionHandler {

    private final HttpHeaders headers = new HttpHeaders();

    @ResponseBody
    @ExceptionHandler({NotFoundException.class})
    @ResponseStatus(HttpStatus.NOT_FOUND)
    ResponseEntity<Object> objectNotFoundHandler(NotFoundException ex, WebRequest request) {
        logException(ex);
        return handleExceptionInternal(ex, null, headers, HttpStatus.NOT_FOUND, request);
    }

    @ResponseBody
    @ExceptionHandler({EmptyResultDataAccessException.class})
    @ResponseStatus(HttpStatus.NO_CONTENT)
    ResponseEntity<Object> objectNotFoundHandler(EmptyResultDataAccessException ex, WebRequest request) {
        logException(ex);
        return handleExceptionInternal(ex, null, headers, HttpStatus.NO_CONTENT, request);
    }

    @ResponseBody
    @ExceptionHandler({AccessDeniedException.class})
    @ResponseStatus(HttpStatus.FORBIDDEN)
    ResponseEntity<Object> forbidden(AccessDeniedException ex, WebRequest request) {
        logException(ex);
        return handleExceptionInternal(ex, ex.getCause(), headers, HttpStatus.FORBIDDEN, request);
    }

    //TODO NoSuchElementException

    private void logException(RuntimeException ex) {
        logger.info("handleException");
        logger.error(ex.getMessage(), ex);
    }

}

File: src/main/kotlin/br/com/controle/financeiro/controller/RestResponseEntityExceptionHandler.kt
package br.com.controle.financeiro.controller

import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
import org.springframework.web.context.request.WebRequest
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.dao.EmptyResultDataAccessException
import br.com.controle.financeiro.model.exception.NotFoundException

@RestControllerAdvice
class RestResponseEntityExceptionHandler : ResponseEntityExceptionHandler() {
    private val headers = HttpHeaders()

    @ResponseBody
    @ExceptionHandler(NotFoundException::class)
    @ResponseStatus(HttpStatus.NOT_FOUND)
    fun objectNotFoundHandler(ex: NotFoundException, request: WebRequest): ResponseEntity<Any>? {
        logException(ex)
        return handleExceptionInternal(ex, null, headers, HttpStatus.NOT_FOUND, request)
    }

    @ResponseBody
    @ExceptionHandler(EmptyResultDataAccessException::class)
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun objectNotFoundHandler(ex: EmptyResultDataAccessException, request: WebRequest): ResponseEntity<Any>? {
        logException(ex)
        return handleExceptionInternal(ex, null, headers, HttpStatus.NO_CONTENT, request)
    }

    //TODO NoSuchElementException
    private fun logException(ex: RuntimeException) {
        logger.info("handleException")
        logger.error(ex.message, ex)
    }
}
